example/diffusion_gmg/diffusion_equation.hpp

```diff
@@ -0,0 +1,320 @@
+  const int ndim = md->GetMeshPointer()->ndim;
+  using TE = parthenon::TopologicalElement;
+  TE te = TE::CC;
+  IndexRange ib = md->GetBoundsI(IndexDomain::interior, te);
+  IndexRange jb = md->GetBoundsJ(IndexDomain::interior, te);
+  IndexRange kb = md->GetBoundsK(IndexDomain::interior, te);
+
+  auto pkg = md->GetMeshPointer()->packages.Get("diffusion_package");
+  const auto alpha = pkg->Param<Real>("diagonal_alpha");
+
+  int nblocks = md->NumBlocks();
+  std::vector<bool> include_block(nblocks, true);
+
+  static auto desc =
+      parthenon::MakePackDescriptor<var_t>(md.get(), {}, {PDOpt::WithFluxes});
+  static auto desc_out = parthenon::MakePackDescriptor<var_t>(md_out.get());
+  auto pack = desc.GetPack(md.get(), include_block);
+  auto pack_out = desc_out.GetPack(md_out.get(), include_block);
+  parthenon::par_for(
+      "FluxMultiplyMatrix", 0, pack.GetNBlocks() - 1, kb.s, kb.e, jb.s, jb.e, ib.s, ib.e,
+      KOKKOS_LAMBDA(const int b, const int k, const int j, const int i) {
+        const auto &coords = pack.GetCoordinates(b);
+        Real dx1 = coords.template Dxc<X1DIR>(k, j, i);
+        pack_out(b, te, var_t(), k, j, i) = -alpha * pack(b, te, var_t(), k, j, i);
+        pack_out(b, te, var_t(), k, j, i) += (1.0) *
+                                             (pack.flux(b, X1DIR, var_t(), k, j, i) -
+                                              pack.flux(b, X1DIR, var_t(), k, j, i + 1)) /
+                                             dx1;
+
+        if (ndim > 1) {
+          Real dx2 = coords.template Dxc<X2DIR>(k, j, i);
+          pack_out(b, te, var_t(), k, j, i) +=
+              (1.0) *
+              (pack.flux(b, X2DIR, var_t(), k, j, i) -
+               pack.flux(b, X2DIR, var_t(), k, j + 1, i)) /
+              dx2;
+        }
+
+        if (ndim > 2) {
+          Real dx3 = coords.template Dxc<X3DIR>(k, j, i);
+          pack_out(b, te, var_t(), k, j, i) +=
+              (pack.flux(b, X3DIR, var_t(), k, j, i) -
+               pack.flux(b, X3DIR, var_t(), k + 1, j, i)) /
+              dx3;
+        }
+      });
+  return TaskStatus::complete;
+}
+
+// This class implement methods for calculating A.x = y and returning the diagonal of A,
+// where A is the the matrix representing the discretized Diffusion equation on the grid.
+// Here we implement the Laplace operator in terms of a flux divergence to (potentially)
+// consistently deal with coarse fine boundaries on the grid. Only the routines Ax and
+// SetDiagonal need to be defined for interfacing this with solvers. The other methods
+// are internal, but can't be marked private or protected because they launch kernels
+// on device.
+template <class var_t, class D_t>
+class DiffusionEquation {
+ public:
+  bool do_flux_cor = false;
+  bool set_flux_boundary = false;
+  bool include_flux_dx = false;
+
+  using IndependentVars = parthenon::TypeList<var_t>;
+
+  DiffusionEquation(parthenon::ParameterInput *pin, const std::string &label) {
```

Comment:
I *think* you could set all these variables in an initializer list, rather than the constructor body. Slightly more modern C++ism.

Reply:
Done